chain/src/chain_service.rs

```diff
@@ -0,0 +1,156 @@
+                        // asynchronous_process_block doesn't interact with tx-pool,
+                        // no need to pause tx-pool's chunk_process here.
+                        let _trace_now = minstant::Instant::now();
+                        self.asynchronous_process_block(lonely_block);
+                        if let Some(handle) = ckb_metrics::handle(){
+                            handle.ckb_chain_async_process_block_duration.observe(_trace_now.elapsed().as_secs_f64())
+                        }
+                        let _ = responder.send(());
+                    },
+                    _ => {
+                        error!("process_block_receiver closed");
+                        break;
+                    },
+                },
+                recv(clean_expired_orphan_timer) -> _ => {
+                    self.orphan_broker.clean_expired_orphans();
+                },
+                recv(signal_receiver) -> _ => {
+                    info!("ChainService received exit signal, exit now");
+                    break;
+                }
+            }
+        }
+    }
+
+    fn non_contextual_verify(&self, block: &BlockView) -> Result<(), Error> {
+        let consensus = self.shared.consensus();
+        BlockVerifier::new(consensus).verify(block).map_err(|e| {
+            debug!("[process_block] BlockVerifier error {:?}", e);
+            e
+        })?;
+
+        NonContextualBlockTxsVerifier::new(consensus)
+            .verify(block)
+            .map_err(|e| {
+                debug!(
+                    "[process_block] NonContextualBlockTxsVerifier error {:?}",
+                    e
+                );
+                e
+            })
+            .map(|_| ())
+    }
+
+    // `self.non_contextual_verify` is very fast.
+    fn asynchronous_process_block(&self, lonely_block: LonelyBlock) {
+        let block_number = lonely_block.block().number();
+        let block_hash = lonely_block.block().hash();
+        // Skip verifying a genesis block if its hash is equal to our genesis hash,
+        // otherwise, return error and ban peer.
+        if block_number < 1 {
+            if self.shared.genesis_hash() != block_hash {
+                warn!(
+                    "receive 0 number block: 0-{}, expect genesis hash: {}",
+                    block_hash,
+                    self.shared.genesis_hash()
+                );
+                self.shared
+                    .insert_block_status(lonely_block.block().hash(), BlockStatus::BLOCK_INVALID);
+                let error = InternalErrorKind::System
+                    .other("Invalid genesis block received")
+                    .into();
+                lonely_block.execute_callback(Err(error));
+            } else {
+                warn!("receive 0 number block: 0-{}", block_hash);
+                lonely_block.execute_callback(Ok(false));
+            }
+            return;
+        }
+
+        if lonely_block.switch().is_none()
```

Comment:
since `switch` is only used here, what's the difference between switch is `Some(None)` and `Switch::None`, maybe we can just set the `switch` field an enum without option in `LonlyBlock`:

```rust
pub struct LonglyBlock {
   ...
   pub switch: Switch
}
```

Reply:
Switch::None represents that all validation rules in this block will not be ignored.

Some(None): indicates that this block does not explicitly specify "how to validate this block," and the default will be `Switch::None`.